keras_contrib/callbacks/dead_relu_detector.py

```diff
@@ -44,17 +45,41 @@ def get_relu_activations(self):
         layer_outputs = [func(list_inputs)[0] for func in funcs]
         for layer_index, layer_activations in enumerate(layer_outputs):
             if self.is_relu_layer(self.model.layers[layer_index]):
-                yield [layer_index, layer_activations]
+                layer_name = self.model.layers[layer_index].name
+                # layer_weight is a list [W] (+ [b])
+                layer_weight = self.model.layers[layer_index].get_weights()
+                # with kernel and bias, the weights are saved as a list [W, b]. If only weights, it is [W]
+                assert type(layer_weight) == list
+                layer_weight_shape = np.shape(layer_weight[0])
+                yield [layer_index, layer_activations, layer_name, layer_weight_shape]
 
     def on_epoch_end(self, epoch, logs={}):
         for relu_activation in self.get_relu_activations():
-            layer_index, activation_values = relu_activation
-            total_neurons = activation_values.shape[-1]
-            dead_neurons = np.sum(activation_values == 0)
-            dead_neurons_share = dead_neurons / total_neurons
+            layer_index, activation_values, layer_name, layer_weight_shape = relu_activation
+
+            shape_act = activation_values.shape
+
+            weight_len = len(layer_weight_shape)
+            act_len = len(shape_act)
+
+            # should work for both Conv and Flat
+            if K.image_data_format() == 'channels_last':
+                # features in last axis
+                axis_filter = -1
+            else:
+                # features before the convolution axis, for weight_len the input and output have to be subtracted
+                axis_filter = -1 - (weight_len - 2)
+
+            total_featuremaps = shape_act[axis_filter]
+
+            axis = tuple(
+                i for i in range(act_len) if (i != axis_filter) and (i != (len(shape_act) + axis_filter)))
+
+            dead_neurons = np.sum(np.sum(activation_values, axis=axis) == 0)
+
+            dead_neurons_share = dead_neurons / total_featuremaps
             if (self.verbose and dead_neurons > 0) or dead_neurons_share > self.dead_neurons_share_threshold:
-                warnings.warn(
-                    'Layer #{} has {} dead neurons ({:.2%})!'
-                        .format(layer_index, dead_neurons, dead_neurons_share),
-                    RuntimeWarning
-                )
+                str_warning = 'Layer {} (#{}) has {} dead neurons ({:.2%})!'.format(layer_name, layer_index,
+                                                                                    dead_neurons, dead_neurons_share)
+                print(str_warning)
```

Comment:
should this print be here? make output matches how other keras calls handle it

Reply:
I put the print there together with the warning since the warning stopped showing after some 2 epochs of dead neurons. Shall I just remove the print? It might have something to do with just my settings for warnings?
" "default" | print the first occurrence of matching warnings for each location where the warning is issued " https://docs.python.org/2/library/warnings.html#the-warnings-filter

Reply:
@ahundt you mean this should be logged to stderr? Makes sense IMHO

Reply:
Okey, I fixed the problem at my side. The warning now shows everytime by setting:
>>>import warnings
>>>warnings.simplefilter("always")
or in terminal by : python python -W alwys foo.py

Now we only have to descide if we show a warning or just print it.

Reply:
Or we give it as a setting:

>def foo(bool_warning = False):
>    if bool_warning:
>        warnings.warn(str_warning, RuntimeWarning)
>    else:
>        print(str_warning)

Reply:
Is warnings imported by keras? I'm just saying to double check how similar APIs in keras do things, maybe they just print most of the time then it is fine.

Reply:
Yes, keras importants warnings in keras/callbacks.py. I think warnings is a standard library?

In my opinion a print will be saver. A warning is nicer. So I don't know if we should give the audience the choice between a print and a warning. I think this should be a decision from higher up (for example you :) )
Although if I look at other examples that use warnings. Here the warning isn't that crucial (unless ALL neurons are dead).

Reply:
Yes I looked at keras/callbacks.py, and based on that I think just a print would be fine.